presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/IntermediateAggregationOptimizer.java

```diff
@@ -0,0 +1,335 @@
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.optimizations;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.SystemSessionProperties;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.planner.DependencyExtractor;
+import com.facebook.presto.sql.planner.Partitioning;
+import com.facebook.presto.sql.planner.PartitioningScheme;
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.ChildReplacer;
+import com.facebook.presto.sql.planner.plan.ExchangeNode;
+import com.facebook.presto.sql.planner.plan.IndexJoinNode;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.PlanRewriter;
+import com.facebook.presto.sql.planner.plan.SemiJoinNode;
+import com.facebook.presto.sql.planner.plan.TableScanNode;
+import com.facebook.presto.sql.planner.plan.UnionNode;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.ExpressionRewriter;
+import com.facebook.presto.sql.tree.ExpressionTreeRewriter;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.SymbolReference;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+
+import static com.facebook.presto.sql.planner.SystemPartitioningHandle.FIXED_ARBITRARY_DISTRIBUTION;
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Step.FINAL;
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Step.INTERMEDIATE;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Scope.REMOTE;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.GATHER;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.REPARTITION;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Inserts an INTERMEDIATE aggregation step with a fixed distribution between a PARTIAL aggregation that
+ * is source partitioned and a FINAL aggregation that occurs on a single node.
+ * <p>
+ * From:
+ * <pre>
+ * - Aggregation (FINAL)
+ *   - Exchange (GATHER)
+ *     - Aggregation (PARTIAL) - same fragment as TableScan
+ * </pre>
+ * to:
+ * <pre>
+ * - Aggregation (FINAL)
+ *   - Exchange (GATHER)
+ *     - Aggregation (INTERMEDIATE)
+ *       - EXCHANGE (ARBITRARY)
+ *         - Aggregation (PARTIAL) - same fragment as TableScan
```

Comment:
Ok, so you are saying that there is a difference whether `AggregateFinal` aggregates 16 * 1000 partial inputs or 16 * 100 partial inputs? Is it THAT big a deal? in such case, wouldn't it be better to create `AggregateIntermediate` after remote gathering exchange and place a local gathering exchange between `AggregateFinal` and `AggregateIntermediate`?

Besides, keep in mind that IMO my proposal would be more general and better in general case, while still significantly decreasing amount of work done by `AggregateFinal` from `1000 000 000` partial inputs down to `16 * 1000` partial inputs`.